minimal_data_sharing/minimal_atomic_message.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/u_int32.hpp"
+
+using namespace std::chrono_literals;
+
+static_assert(std::atomic<std_msgs::msg::UInt32>::is_always_lock_free);
+
+// Bigger data types are not lock free, this asserts would fail
+// static_assert(std::atomic<std_msgs::msg::String>::is_always_lock_free);
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), msg_(std_msgs::msg::UInt32())
+  {
+    // Make the callback groups reentrant so they can run concurrently
+    auto reentrant_callback_group =
+      this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
+    rclcpp::PublisherOptions options;
+    options.callback_group = reentrant_callback_group;
+    for (auto pub_index = 0; pub_index < 10; pub_index++) {
+      auto publisher = this->create_publisher<std_msgs::msg::UInt32>("topic", 10, options);
+      auto timer_callback =
+        [this, pub_index, publisher]() -> void {
+          auto msg = increment_message_data();
+          RCLCPP_INFO(this->get_logger(), "Publisher %d: '%d'", pub_index, msg.data);
+          publisher->publish(msg);
+        };
+      timers_.push_back(this->create_wall_timer(100ms, timer_callback, reentrant_callback_group));
+    }
+  }
+
+  std_msgs::msg::UInt32 increment_message_data()
+  {
+    auto old_value = msg_.load();
+    auto new_value = old_value;
+    new_value.data++;
+    do {
+      if (msg_.compare_exchange_strong(old_value, new_value)) {
+        break;
+      }
+    } while(true);
```

Comment:
Can this be

```suggestion
    do {
    } while(!msg_.compare_exchange_strong(old_value, new_value));
```

Also, since `compare_and_exchange` will store the value to `old_value` if it fails, should this code be the following (move the `new_value` calculation into the loop)? I assume you basically want an atomic counter shared between multiple threads? The way above seems to be prone to a race condition? Or maybe I'm mis-understanding something.

```suggestion
    do {
      new_value = old_value;
      new_value.data++;
    } while(!msg_.compare_exchange_strong(old_value, new_value));
```

Also, I'm pretty sure you can get away with the weak version, as per the docs:

> The weak forms (1-2) of the functions are allowed to fail spuriously, that is, act as if `*this != expected` even if they are equal. When a compare-and-exchange is in a loop, the weak version will yield better performance on some platforms. 

So:

```suggestion
    do {
      new_value = old_value;
      new_value.data++;
    } while(!msg_.compare_exchange_weak(old_value, new_value));
```

Am I misunderstanding something...?

Reply:
I would like to wait for @MatthiasKillat input here. 

In any case, I'm considering removing this example for two reasons:
- The example is probably not a good case. I would rather use a more complex message type with more fields and show how to update the whole message atomically.
- One of the conclusions for this package should be: don't create your own lock-free algorithms. It's better to rely upon widely used libraries and well-tested code because is very error-prone. This example might send the wrong message (no pun intended).

Reply:
Yeah, you're right. The code was wrong.